### PR TITLE
Fix adding actors to ITIL Objects

### PR DESCRIPTION
**Pull request by cconard96**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #11796

Fix bug that prevented adding any actors to ITIL Objects if you have business rules. For 10.0, this means just Ticket rules.

### CODE REVIEW

**Comment by cmoreno038** [comment]
Changes were made to the mentioned file and more actors and groups were added corre
- Before add actor
imagen
ctly

- Adding actor
imagen

- Saving changes
imagen

-History

imagen

It ran without problems adding and removing users
Regards

**Comment by cconard96** [comment]
This is still a work in progress and based on the failing tests, it probably breaks other stuff. This fix should not be used in production right now.

**Review by cmoreno038 — APPROVED** [review]

**Comment by cconard96** [comment]
I think the last fix here will work OK without side effects. At least, it doesn't break something that has tests.

**Comment by cmoreno038** [comment]
I´m sorry I already backup file and add 6 lines and the portal crashed when save changes by add actors in the ticket

- Add actor
imagen

- Save changes
imagen

- Blank Page
imagen

The tests were done with Edge and Firefox and it is the same result

**Comment by cconard96** [comment]
Please check "files/_log/php-errors.log" to see what the error was.

**Comment by cmoreno038** [comment]
Hi, this is log when page crashed

[2022-06-03 23:19:25] glpiphplog.CRITICAL:   *** Uncaught Exception TypeError: Argument 1 passed to Glpi\Toolbox\Sanitizer::isNsClassOrCallableIdentifier() must be of the type string, array given, called in /var/www/html/glpi/src/DBmysql.php on line 1257 in /var/www/html/glpi/src/Toolbox/Sanitizer.php at line 228
  Backtrace :
  src/DBmysql.php:1257                               Glpi\Toolbox\Sanitizer::isNsClassOrCallableIdentifier()
  src/DBmysql.php:1288                               DBmysql::quoteValue()
  src/DBmysql.php:1312                               DBmysql->buildInsert()
  src/CommonDBTM.php:705                             DBmysql->insert()
  src/CommonDBTM.php:1297                            CommonDBTM->addToDB()
  src/CommonITILObject.php:3012                      CommonDBTM->add()
  src/CommonITILObject.php:1716                      CommonITILObject->addAdditionalActors()
  src/Ticket.php:1303                                CommonITILObject->prepareInputForUpdate()
  src/CommonDBTM.php:1565                            Ticket->prepareInputForUpdate()
  front/ticket.form.php:84                           CommonDBTM->update()
  
[2022-06-03 23:20:02] glpiphplog.WARNING: plugin_init_roundrobin - plugin initialization  
[2022-06-03 23:20:02] glpiphplog.WARNING: PluginRoundRobinConfig::init - defining hooks handlers  
[2022-06-03 23:20:02] glpiphplog.WARNING: PluginRoundRobinConfig::loadSources - loading sources...  
[2022-06-03 23:21:02] glpiphplog.WARNING: plugin_init_roundrobin - plugin initialization  
[2022-06-03 23:21:02] glpiphplog.WARNING: PluginRoundRobinConfig::init - defining hooks handlers  
[2022-06-03 23:21:02] glpiphplog.WARNING: PluginRoundRobinConfig::loadSources - loading sources...  
[2022-06-03 23:22:02] glpiphplog.WARNING: plugin_init_roundrobin - plugin initialization  
[2022-06-03 23:22:02] glpiphplog.WARNING: PluginRoundRobinConfig::init - defining hooks handlers  
[2022-06-03 23:22:02] glpiphplog.WARNING: PluginRoundRobinConfig::loadSources - loading sources...  
[2022-06-03 23:19:25] glpiphplog.CRITICAL:   *** Uncaught Exception TypeError: Argument 1 passed to Glpi\Toolbox\Sanitizer::isNsClassOrCallableIdentifier() must be of the type string, array given, called in /var/www/html/glpi/src/DBmysql.php on line 1257 in /var/www/html/glpi/src/Toolbox/Sanitizer.php at line 228
  Backtrace :
  src/DBmysql.php:1257                               Glpi\Toolbox\Sanitizer::isNsClassOrCallableIdentifier()
  src/DBmysql.php:1288                               DBmysql::quoteValue()
  src/DBmysql.php:1312                               DBmysql->buildInsert()
  src/CommonDBTM.php:705                             DBmysql->insert()
  src/CommonDBTM.php:1297                            CommonDBTM->addToDB()
  src/CommonITILObject.php:3012                      CommonDBTM->add()
  src/CommonITILObject.php:1716                      CommonITILObject->addAdditionalActors()
  src/Ticket.php:1303                                CommonITILObject->prepareInputForUpdate()
  src/CommonDBTM.php:1565                            Ticket->prepareInputForUpdate()
  front/ticket.form.php:84                           CommonDBTM->update()
  
[2022-06-03 23:20:02] glpiphplog.WARNING: plugin_init_roundrobin - plugin initialization  

Regards

**Comment by cmoreno038** [comment]
Hi,

I already deactivated the plugins, and appears same error

[2022-06-03 23:29:46] glpiphplog.CRITICAL:   *** Uncaught Exception TypeError: Argument 1 passed to Glpi\Toolbox\Sanitizer::isNsClassOrCallableIdentifier() must be of the type string, array given, called in /var/www/html/glpi/src/DBmysql.php on line 1257 in /var/www/html/glpi/src/Toolbox/Sanitizer.php at line 228
  Backtrace :
  src/DBmysql.php:1257                               Glpi\Toolbox\Sanitizer::isNsClassOrCallableIdentifier()
  src/DBmysql.php:1288                               DBmysql::quoteValue()
  src/DBmysql.php:1312                               DBmysql->buildInsert()
  src/CommonDBTM.php:705                             DBmysql->insert()
  src/CommonDBTM.php:1297                            CommonDBTM->addToDB()
  src/CommonITILObject.php:3012                      CommonDBTM->add()
  src/CommonITILObject.php:1716                      CommonITILObject->addAdditionalActors()
  src/Ticket.php:1303                                CommonITILObject->prepareInputForUpdate()
  src/CommonDBTM.php:1565                            Ticket->prepareInputForUpdate()
  front/ticket.form.php:84                           CommonDBTM->update()

regards

**Comment by cconard96** [comment]
The commit I made was missing somehow. I have recreated it. Please test again when you have a chance,

**Review by cmoreno038 — APPROVED** [review]
I was restored CommonITILObject.php, add changes in Ticket.php file and test successfuly working, with and without plugins, add and put off actors in tickets and same result ok

**Comment by trasher** [comment]
I wonder if we can add tests on that part

**Comment by cmoreno038** [comment]
- I select the ticket
imagen

-Participant is added
imagen

- Save changes

imagen

File: src/Ticket.php [hunk 1]
```diff
@@ -1259,12 +1259,28 @@ public function prepareInputForUpdate($input)
...
+                                $input[$additionalfield] = [];
+                             }
+                             if (!isset($input[$additionalfield]['users_id'])) {
+                                $input[$additionalfield]['users_id'] = [];
+                             }
+                             if (is_array($input[$k])) {
+                                 $input[$additionalfield]['users_id'] = array_merge($input[$additionalfield]['users_id'] ?? [], $input[$k] ?? []);
+                             } else {
+                                 $input[$additionalfield][] = ['users_id' => $input['_' . $a . 's_id_' . $t]];
+                             }
```
Comment (cedric-anne) [inline src/Ticket.php:1273]: With this, `$input[$additionalfield]['users_id']` may still be both an array or a string. Why not always use an array ?

File: tests/functionnal/Ticket.php [hunk 2]
```diff
@@ -4509,4 +4509,69 @@ public function testNeedReopen()
...
+                        'users_id'  => 1
+                    ],
+                    '_users_id_observer'  => 3,
+                    '_users_id_assign'    => 4,
+                    '_additional_groups_assigns' => [1],
+                    '_groups_id_assign'      => [2],
+                    '_suppliers_id_assign'      => [4],
+                ],
+                'changes' => [
+                    '_additional_requesters' => [ // Does not make sense, but whatever. It works I guess...
```
Comment (cedric-anne) [inline tests/functionnal/Ticket.php:4532]: Makes more sense now.
```suggestion
                    '_additional_requesters' => [
```

**Review by cedric-anne — COMMENTED** [review]
Is it possible to produce a functionnal test that creates an add/update rule, creates a ticket with some actors (mix of users/groups/suppliers, assigned/observers/requesters), and updates this ticket with some other actors in the `_actors` input (mix of users/groups/suppliers, assigned/observers/requesters, added/updated/deleted) ? Using a data provider  for this test could help to be sure that many different cases are tested.

**Comment by cconard96** [comment]
Broken. Do not use patch.

**Pull request closed without merging**